### PR TITLE
feat(pt_expt): add eval_typeebd, eval_descriptor, eval_fitting_last_layer

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Summary

- Add `eval_typeebd`, `eval_descriptor`, `eval_fitting_last_layer` to pt_expt's `DeepEval`, using the eager `_dpmodel` for diagnostic computation
- Add `get_dp_atomic_model()` API to the model hierarchy (make_model, FrozenModel, SpinModel) for clean access to the underlying `DPAtomicModel`
- Add `set_return_middle_output` to `GeneralFitting` with proper `FittingOutputDef` registration — `fitting_check_output` now evaluates `output_def()` dynamically
- Preserve `middle_output` through DipoleFitting/PolarFitting `call()` by modifying the results dict in-place
- Add cross-backend reference values for fparam_aparam model (descriptor + fit_ll)
- Remove .pte/.pt2 skips in `test_models.py` for `test_descriptor` and `test_fitting_last_layer`

## Limitations

- Spin models: `eval_descriptor` and `eval_fitting_last_layer` raise `NotImplementedError` (SpinModel preprocesses coords with virtual atoms)
- DPZBLModel: `get_dp_atomic_model()` returns `None` (LinearAtomicModel has no single descriptor/fitting_net)
- `model_check_output` still caches `output_def` at init (only `fitting_check_output` is dynamic) — not an issue since `middle_output` is consumed at fitting level
- `_middle_output_def()` added to all fitting subclasses but only tested for InvarFitting, DipoleFitting, PolarFitting (not PropertyFitting, DOSFitting)

## Test plan

- [x] 10 dpmodel middle_output tests (shape, toggle, output_def registration, decorator validation, dipole/polar passthrough)
- [x] 6 eval diagnostic tests (typeebd, descriptor, fitting_last_layer for se_e2_a and DPA1)
- [x] 4 get_dp_atomic_model tests (energy, ZBL, spin, frozen)
- [x] 3 spin model diagnostic tests (typeebd works, descriptor/fitting raises)
- [x] 2 ASE neighbor list consistency tests
- [x] Cross-backend consistency via fparam_aparam reference values in test_models.py

## Summary by CodeRabbit

* **New Features**
  * Inference APIs now expose type-embedding, descriptor, and fitting-last-layer outputs; models may return intermediate fitting-network outputs and higher-level fitters propagate them.
  * New public accessor to retrieve an underlying atomic model from wrapped models.

* **Refactor**
  * Consolidated input preparation and evaluation flow; multi-output handling and output-definition validation unified.

* **Tests**
  * Large expansion of unit and integration tests and test data for new outputs, toggles, determinism, and model variants.

### CODE REVIEW

File: deepmd/dpmodel/fitting/general_fitting.py [hunk 1]
```diff
@@ -424,6 +428,30 @@ def get_default_fparam(self) -> list[float] | None:
...
+        """
+        self.eval_return_middle_output = enable
+
+    def _middle_output_def(self) -> list[OutputVariableDef]:
+        """Return extra OutputVariableDefs for middle_output when enabled."""
+        if self.eval_return_middle_output:
+            return [
+                OutputVariableDef(
+                    "middle_output",
+                    [self.neuron[-1]],
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/fitting/general_fitting.py:456]: Fixed. `set_return_middle_output(True)` now raises `ValueError` when `neuron=[]`.

[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/dpmodel/fitting/general_fitting.py:456]: Fixed. Raises `ValueError` instead of silently skipping.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 2]
```diff
@@ -955,3 +993,177 @@ def get_model(self) -> torch.nn.Module:
...
         """
         return self.exported_module
+
+    def _is_spin_model(self) -> bool:
+        """Check if the underlying dpmodel is a SpinModel."""
+        from deepmd.dpmodel.model.spin_model import (
+            SpinModel,
+        )
+
+        return isinstance(self._dpmodel, SpinModel)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:1003]: Not changed. `pt_expt.SpinModel` inherits from `dpmodel.SpinModel`, so the `isinstance` check works correctly for both.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 3]
```diff
@@ -955,3 +993,177 @@ def get_model(self) -> torch.nn.Module:
...
+            fparam_for_des = (
+                fparam_t if getattr(dp_am, "add_chg_spin_ebd", False) else None
+            )
+            descriptor, rot_mat, g2, h2, sw = dp_am.descriptor(
+                ext_coord_t,
+                ext_atype_t,
+                nlist_t,
+                mapping=mapping_t,
+                fparam=fparam_for_des,
+            )
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:1095]: Fixed. Unused variables are now underscore-prefixed or use `*_`.

File: source/tests/common/dpmodel/test_fitting_middle_output.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        aparam = np.zeros([nf, nloc, 3], dtype=np.float64)
+        ft.set_return_middle_output(True)
+        ret = ft.call(descriptor, atype, fparam=fparam, aparam=aparam)
+        self.assertIn("middle_output", ret)
+        expected_shape = (nf, nloc, ft.neuron[-1])
+        self.assertEqual(ret["middle_output"].shape, expected_shape)
+
+    def test_middle_output_registered_in_output_def(self) -> None:
+        """middle_output should appear in output_def when enabled."""
+        ft, descriptor, atype = self._build_fitting(mixed_types=True)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/common/dpmodel/test_fitting_middle_output.py]: Fixed. Changed to `ft, _, _ = self._build_fitting(...)`.

File: source/tests/pt_expt/infer/test_deep_eval.py [hunk 5]
```diff
@@ -1299,5 +1300,618 @@ def test_pt2_vs_pte_consistency(self) -> None:
...
+        dp_am = DPAtomicModelDP(ds, ft, type_map=["foo", "bar"])
+        pair_tab = PairTabAtomicModel(
+            tab_file=None, rcut=4.0, sel=14, type_map=["foo", "bar"]
+        )
+        zbl_am = DPZBLLinearEnergyAtomicModel(
+            dp_am, pair_tab, sw_rmin=1.0, sw_rmax=2.0, type_map=["foo", "bar"]
+        )
+        # zbl_am is a LinearEnergyAtomicModel, not DPAtomicModel
+        self.assertFalse(isinstance(zbl_am, DPAtomicModelDP))
+
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval.py:1711]: Not changed. `get_dp_atomic_model()` is a model-level API, not available on atomic models. The isinstance check is correct here.

File: source/tests/pt_expt/infer/test_deep_eval.py [hunk 6]
```diff
@@ -1299,5 +1300,618 @@ def test_pt2_vs_pte_consistency(self) -> None:
...
+        tmpdir = tempfile.mkdtemp()
+        try:
+            pte_path = os.path.join(tmpdir, "test.pte")
+            deserialize_to_file(pte_path, {"model": model.serialize()})
+            dp = DeepPot(pte_path)
+            # The _dpmodel deserialized from .pte is a regular energy model
+            dp_am = dp.deep_eval._dpmodel.get_dp_atomic_model()
+            self.assertIsNotNone(dp_am)
+            self.assertIsInstance(dp_am, DPAtomicModelDP)
+        finally:
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval.py]: Renamed to `test_deserialized_model_delegates` to match what it actually tests.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 7]
```diff
@@ -955,3 +993,173 @@ def get_model(self) -> torch.nn.Module:
...
+                    descriptor,
+                    atype,
+                    gr=rot_mat,
+                    g2=g2,
+                    h2=h2,
+                    fparam=fparam_t,
+                    aparam=aparam_t,
+                )
+            finally:
+                fitting_net.set_return_middle_output(False)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:1176]: Not changed. The `_dpmodel` is a private eager model only used for diagnostics. The try/finally already ensures reset.

File: source/tests/pt_expt/infer/test_deep_eval.py [hunk 8]
```diff
@@ -1299,5 +1300,616 @@ def test_pt2_vs_pte_consistency(self) -> None:
...
+
+        spin = Spin(use_spin=[True, False], virtual_scale=[0.5, 0.0])
+        cls.spin_model = SpinModel(backbone_model=backbone, spin=spin)
+
+        # Export backbone as .pte, then swap _dpmodel to SpinModel
+        cls._tmpdir = tempfile.mkdtemp()
+        pte_path = os.path.join(cls._tmpdir, "spin.pte")
+        deserialize_to_file(pte_path, {"model": backbone.serialize()})
+        cls.dp = DeepPot(pte_path)
+        cls.dp.deep_eval._dpmodel = cls.spin_model
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval.py:1802]: Not applicable. `_is_spin_model()` uses `isinstance(self._dpmodel, SpinModel)`, not `self._is_spin`. Since we swap `_dpmodel` to a `SpinModel` instance, isinstance works correctly.

File: deepmd/pt_expt/infer/deep_eval.py [hunk 9]
```diff
@@ -955,3 +993,177 @@ def get_model(self) -> torch.nn.Module:
...
+
+    def eval_descriptor(
+        self,
+        coords: np.ndarray,
+        cells: np.ndarray | None,
+        atom_types: np.ndarray,
+        fparam: np.ndarray | None = None,
+        aparam: np.ndarray | None = None,
+        **kwargs: Any,
+    ) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply deepmd/pt_expt/infer/deep_eval.py:1040]: Fixed in 4f4983d6. Added `np.array()` coercion for `coords`, `cells`, `atom_types` in both `eval_descriptor` and `eval_fitting_last_layer`, matching `eval()`.

File: source/tests/pt_expt/infer/test_deep_eval.py [hunk 10]
```diff
@@ -1299,5 +1300,636 @@ def test_pt2_vs_pte_consistency(self) -> None:
...
+
+        shutil.rmtree(cls._tmpdir, ignore_errors=True)
+
+    def test_typeebd_dpa1(self) -> None:
+        """DPA1 model has type embedding, should return valid array."""
+        typeebd = self.dp_dpa1.deep_eval.eval_typeebd()
+        self.assertEqual(typeebd.ndim, 2)
+        # DPA1 TypeEmbedNet outputs (ntypes+1) rows (padding type included)
+        self.assertIn(typeebd.shape[0], (self.nt, self.nt + 1))
+        self.assertGreater(typeebd.shape[1], 0)
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/tests/pt_expt/infer/test_deep_eval.py:1367]: Not changed. TypeEmbedNet includes a padding row for the virtual type, so the output is `(ntypes+1, tebd_dim)`. The test intentionally accepts both `ntypes` and `ntypes+1` because this depends on the TypeEmbedNet configuration (`padding=True/False`).

**Review by njzjz — APPROVED** [review]
